File: praticas_r/01-brasilapi.R
library(httr)
library(httr2)
library(jsonlite)
library(tidyverse)

# cep ---------------------------------------------------------------------

u_base <- "https://brasilapi.com.br/api"
endpoint_cep <- "/cep/v1/"

# vamos gerar um CEP válido aqui: https://www.geradordecep.com.br
cep <- "58084435"

u_cep <- paste0(u_base, endpoint_cep, cep)
r_cep <- GET(u_cep)
r_cep

# Vamos entender as diferentes formas de pegar o resultado

content(r_cep)
content(r_cep, as = "text")
content(r_cep, as = "raw")
content(r_cep, as = "parsed")

# usando httr2
req <- u_cep |>
  request()

## alternativa com req_url_path_append()
# req <- u_base |>
#   request() |>
#   req_url_path_append(endpoint_cep) |>
#   req_url_path_append(cep)

resp <- req |>
  req_perform()

r_cep_httr2 |>
  resp_body_json()

r_cep_httr2 |>
  resp_body_string()

# agora vamos pesquisar na tabela FIPE

endpoint_fipe <- "/fipe/marcas/v1/"
tipo_veiculo <- "carros"
u_fipe <- paste0(u_base, endpoint_fipe, tipo_veiculo)
r_fipe <- GET(u_fipe)

## transformando resultado em um data frame

da <- content(r_fipe, simplifyDataFrame = TRUE) |>
  as_tibble()

da

## lendo com jsonlite

content(r_fipe, "text") |>
  fromJSON() |>
  as_tibble()

## salvando em arquivo

r_fipe <- GET(u_fipe, write_disk("fipe.json"))

## lendo arquivo com jsonlite

read_json("fipe.json", simplifyDataFrame = TRUE) |>
  as_tibble()

## e o parâmetro?

endpoint_fipe_tabelas <- "/fipe/tabelas/v1"
u_fipe_tabelas <- paste0(u_base, endpoint_fipe_tabelas)
r_fipe_tabelas <- GET(u_fipe_tabelas)

tabelas <- r_fipe_tabelas |>
  content(simplifyDataFrame = TRUE) |>
  as_tibble()

## chamando com um parâmetro

query_fipe <- list(
  tabela_referencia = "270"
)

r_fipe_query <- GET(u_fipe, query = query_fipe)

## outra forma de usar query

# com httr2

req <- u_base |>
  request() |>
  req_url_path_append(endpoint_fipe_tabelas) |>
  req_url_query(tabela_referencia = "270")

## alternativa
# req <- u_base |>
#   request() |>
#   req_url_path_append(endpoint_fipe_tabelas) |>
#   req_url_query(!!!query_fipe)

resp <- req |>
  req_perform()

resp |>
  resp_body_json(simplifyDataFrame = TRUE) |>
  as_tibble()

# preco de carro na FIPE --------------------------------------------------

## infelizmente ainda não dá para pegar a lista de carros pela Brasil API
## https://github.com/BrasilAPI/BrasilAPI/issues/373

## Mas nós podemos pegar aqui: https://www.tabelafipebrasil.com/fipe/carros
## (depois podemos fazer isso via web scraping!)

endpoint_preco <- "/fipe/preco/v1/"
cod_veiculo <- "004515-2"
u_preco <- paste0(u_base, endpoint_preco, cod_veiculo)
r_preco <- GET(u_preco)

content(r_preco, simplifyDataFrame = TRUE) |>
  as_tibble()

## agora vamos ver o preço do carro em outra tabela

req <- u_base |>
  request() |>
  req_url_path_append(endpoint_preco) |>
  req_url_path_append(cod_veiculo) |>
  req_url_query(tabela_referencia = "270")

resp <- req |>
  req_perform()

resp |>
  resp_body_json(simplifyDataFrame = TRUE) |>
  as_tibble()

## aplicação: analisar o preço do carro 0km em cada mês entre 2022 e 2024

tabelas_filtrado <- tabelas |>
  filter(codigo >= 281)

tabelas_filtrado

pegar_tabela_codigos <- function(tab) {
  # fazendo tudo em um pipeline só
  req |>
    req_url_query(tabela_referencia = tab) |>
    req_perform() |>
    resp_body_json(simplifyDataFrame = TRUE) |>
    as_tibble() |>
    filter(anoModelo == 32000)
}

da_preco <- tabelas_filtrado |>
  pull(codigo) |>
  map(pegar_tabela_codigos) |>
  list_rbind()

File: praticas_py/01-brasilapi.py
# %%
import requests
import pandas as pd
import json

# cep ----------------------------------------------------------------

# %%
u_base = 'https://brasilapi.com.br/api'
endpoint_cep = '/cep/v1/'

# vamos gerar um CEP válido aqui: https://www.geradordecep.com.br
cep = '89010025'

u_cep = u_base + endpoint_cep + cep
r_cep = requests.get(u_cep)
r_cep

# %%
print(r_cep.text)

print(r_cep.json())

print(r_cep.content)

# agora vamos pesquisar na tabela FIPE ---------------------------------------

# %%

endpoint_fipe = '/fipe/marcas/v1/'
tipo_veiculo = 'carros'

u_fipe = u_base + endpoint_fipe + tipo_veiculo

r_fipe = requests.get(u_fipe)

## transfomando resultado em um DataFrame
# %%
df = pd.DataFrame(r_fipe.json())
df

## salvando em um arquivo json
# %%
r_fipe = requests.get(u_fipe)
with open('fipe.json', 'wb') as f:
  f.write(r_fipe.content)

## salvando em um arquivo csv
# %%
df.to_csv('fipe.csv', index=False)

## e o parâmetro?

# %%
endpoint_fipe_tabelas = '/fipe/tabelas/v1'
u_fipe_tabelas = u_base + endpoint_fipe_tabelas
r_fipe_tabelas = requests.get(u_fipe_tabelas)

tabelas = pd.DataFrame(r_fipe_tabelas.json())
tabelas

## chamando com um parâmetro

query_fipe = {'tabela_referencia': 270}

r_fipe = requests.get(u_fipe, params=query_fipe)

pd.DataFrame(r_fipe.json())

# preço de um carro específico --------------------------------------------

## infelizmente ainda não dá para pegar a lista de carros pela Brasil API
## https://github.com/BrasilAPI/BrasilAPI/issues/373

## Mas nós podemos pegar aqui: https://www.tabelafipebrasil.com/fipe/carros

# %%
endpoint_preco = '/fipe/preco/v1/'
cod_veiculo = '004515-2'
u_preco = u_base + endpoint_preco + cod_veiculo

r_preco = requests.get(u_preco)

pd.DataFrame(r_preco.json())

## agora vamos ver o preço do carro em outra tabela

query_preco = {'tabela_referencia': 270}

r_preco = requests.get(u_preco, params=query_preco)

pd.DataFrame(r_preco.json())

## aplicação: analisar o preço do carro 0km em cada mês entre 2022 e 2024

# %%

tabelas_filtrado = tabelas[tabelas['codigo'] >= 281]
tabelas_filtrado

# %%
df_lista = []

for tab in tabelas_filtrado['codigo']:
  query_preco = {'tabela_referencia': tab}
  r_preco = requests.get(u_preco, params=query_preco)
  df_preco = pd.DataFrame(r_preco.json()).query('anoModelo == 32000')
  df_preco['tabela_referencia'] = tab
  df_lista.append(df_preco)

df_preco = pd.concat(df_lista)
df_preco